test/run_test.py

```diff
@@ -1161,36 +1167,66 @@ def query_changed_test_files() -> List[str]:
     lines = [line.strip() for line in lines]
     return lines
 
+def query_failure_test_module(reports: List[Tuple["Report", str]]) -> List[str]:
+    test_modules = []
+    if len(reports) == 0 or len(reports[0]) == 0:
+        return test_modules
+    report = reports[0][0]
+    assert report.get('format_version') == 2, "S3 format currently handled is version 2 only"
+    files: Dict[str, Any] = report['files']
+    for fname, file in files.items():
+        contains_failure = any(
+            any(case['status'] == 'errored' or case['status'] == 'failed'
+                for _, case in suite['cases'].items())
+            for _, suite in file['suites'].items())
+        if contains_failure:
+            test_modules.append(fname)
+    return test_modules
+
 
 def reorder_tests(tests: List[str]) -> List[str]:
-    try:
-        changed_files = query_changed_test_files()
-    except Exception:
-        # If unable to get changed files from git, quit without doing any sorting
-        return tests
+    prioritized_tests = []
+    # Try using historic stats from PR.
+    if ENABLE_PR_HISTORY_REORDERING and HAVE_BOTO3:
+        pr_number = os.environ.get("CIRCLE_PR_NUMBER", "")
+        if len(pr_number):
+            ci_job_prefix = get_stripped_CI_job()
+            s3_reports: List[Tuple["Report", str]] = get_previous_reports_for_pr(
+                pr_number, ci_job_prefix)
+            prioritized_tests = query_failure_test_module(s3_reports)
+            print("Prioritized test from previous CI info.")
+
+    # Using file changes priority if no stats found from previous PR.
+    if len(prioritized_tests) == 0:
+        try:
+            changed_files = query_changed_test_files()
+        except Exception:
+            # If unable to get changed files from git, quit without doing any sorting
+            return tests
 
-    prefix = f"test{os.path.sep}"
-    changed_tests = [f for f in changed_files if f.startswith(prefix) and f.endswith(".py")]
-    changed_tests = [f[len(prefix):] for f in changed_tests]
-    changed_tests = [f[:-len(".py")] for f in changed_tests]
+        prefix = f"test{os.path.sep}"
+        prioritized_tests = [f for f in changed_files if f.startswith(prefix) and f.endswith(".py")]
+        prioritized_tests = [f[len(prefix):] for f in prioritized_tests]
+        prioritized_tests = [f[:-len(".py")] for f in prioritized_tests]
+        print("Prioritized test from test file changes.")
 
     bring_to_front = []
     the_rest = []
 
     for test in tests:
-        if test in changed_tests:
+        if test in prioritized_tests:
             bring_to_front.append(test)
         else:
             the_rest.append(test)
-
-    sorted_tests = bring_to_front + the_rest
-
-    if len(sorted_tests) != len(tests):
-        # Something went wrong, bail out without doing any sorting
+    if len(tests) == len(bring_to_front) + len(the_rest):
```

Comment:
something like this might be more robust but it's up to you:
```suggestion
    if sorted(tests) == sorted(bring_to_front + the_rest):
```

Reply:
it doesn't really matters as it can't create tests if it is not already in the `tests` collection.

Reply:
but the whole point of this check is to catch unexpected bugs lol